File: csharp/lesson 3/SubarraySum.cs
using System;
using System.Collections.Generic;


// https://leetcode.com/problems/subarray-sum-equals-k/
public class Solution 
{
    public int SubarraySum(int[] nums, int k) 
    {
        var sums = new Dictionary<int, int>() { {0, 1} };
        
        int result = 0, sum = 0;
        foreach (var num in nums) 
        {            
            sum += num;
            result += sums.GetValueOrDefault(sum - k, 0);
            sums[sum] = sums.GetValueOrDefault(sum, 0) + 1;
        }
        
        return result;
    }
}

File: javascript/lesson 3/subarraySum.js
// https://leetcode.com/problems/subarray-sum-equals-k/
function subarraySum(nums, k) {
    const sums = new Map([[0, 1]]);
    let result = 0, sum = 0;
    nums.forEach(num => {
        sum += num;
        result += sums.get(sum - k) || 0;
        sums.set(sum, (sums.get(sum) || 0) + 1);
    });
    return result;
}
